File: python/questions/productExceptSelf.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n 
        pp = 1
        for i in range(n):
            ans[i] = pp
            pp *= nums[i]
        pp = 1
        for i in range(n-1, -1, -1):
            ans[i] *= pp
            pp *= nums[i]
        return ans

File: typescript/questions/productExceptSelf.ts
function productExceptSelf(nums: number[]): number[] {
  const nl = nums.length;
  let ans: number[] = new Array(nl);    
  ans[nl-1] = 1;
  for (let p=nl-2; p>=0; --p) {
    ans[p] = ans[p+1] * nums[p+1];
  }
  let pre = nums[0];
  for (let p=1; p<nl; ++p) {
    ans[p] *= pre;
    pre *= nums[p];
  }
  return ans;
};
